daml-lf/data/src/main/scala/com/digitalasset/daml/lf/data/IdString.scala

```diff
@@ -0,0 +1,257 @@
+// Copyright (c) 2020 The DAML Authors. All rights reserved.
+// SPDX-License-Identifier: Apache-2.0
+package com.digitalasset.daml.lf.data
+
+import scalaz.Equal
+
+sealed trait StringModule[T] {
+
+  def fromString(s: String): Either[String, T]
+
+  @throws[IllegalArgumentException]
+  def assertFromString(s: String): T
+
+  implicit def equalInstance: Equal[T]
+
+  implicit def ordering: Ordering[T]
+
+  // We provide the following array factory instead of a ClassTag
+  // because the latter lets people easily reinterpret any string as a T.
+  // See
+  //  * https://github.com/digital-asset/daml/pull/983#discussion_r282513324
+  //  * https://github.com/scala/bug/issues/9565
+  val Array: ArrayFactory[T]
+
+  def toStringMap[V](map: Map[T, V]): Map[String, V]
+}
+
+/** ConcatenableMatchingString are non empty US-ASCII strings built with letters, digits,
+  * and some other (parameterizable) extra characters.
+  * We use them to represent identifiers. In this way, we avoid
+  * empty identifiers, escaping problems, and other similar pitfalls.
+  *
+  * ConcatenableMatchingString has the advantage over MatchingStringModule of being
+  * concatenable and can be generated from numbers without extra checks.
+  * Those properties are heavily use to generate some ids by combining other existing
+  * ids.
+  */
+sealed trait ConcatenableStringModule[T <: String] extends StringModule[T] {
+
+  def encode(a: Array[Byte]): T
+
+  def fromLong(i: Long): T
+
+  def fromInt(i: Int): T
+
+  def concat(s: T, ss: T*): T
+}
+
+sealed trait UnionStringModule[T <: String, TV0 <: T, TV1 <: T] extends StringModule[T] {
```

Comment:
One can clearly see the source of inspiration for the type parameter names ;)
But I'd go for `TA` and `TB` (or `U` and `V`) instead of `TV0` and `TV1` respectively.

Reply:
fixed